File: server_elixir/test/afterlight/activities/fishing_test.exs
defmodule Afterlight.Activities.FishingTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.Fishing

  describe "init_sim_state/1" do
    test "opens a social session with idle lines and no economy fields" do
      state = Fishing.init_sim_state(slots: [0, 1], now_ms: 1_700_000_000_000)

      assert state["status"] == "open"
      assert state["environment"]["policy"] == "live"
      assert state["environmentAt"] == 1_700_000_000_000
      assert state["anglers"]["0"]["phase"] == "idle"
      assert state["anglers"]["1"]["phase"] == "idle"

      blob = inspect(state)
      refute blob =~ "gold"
      refute blob =~ "inventory"
      refute blob =~ "xp"
      refute blob =~ "crop"
    end
  end

  describe "validate_controls/1 and apply_input/3" do
    test "accepts cast/reel/release and rejects unknown kinds" do
      assert {:ok, %{"kind" => "cast", "power" => 1.0}} =
               Fishing.validate_controls(%{"kind" => "cast", "power" => 9})

      assert {:ok, %{"kind" => "reel"}} = Fishing.validate_controls(%{"kind" => "reel"})
      assert {:error, :unknown_kind} = Fishing.validate_controls(%{"kind" => "warp"})
      assert {:error, :invalid_controls} = Fishing.validate_controls("nope")
    end

    test "two visitors each see both lines after independent casts" do
      state = Fishing.init_sim_state(slots: [0, 1], seed: 7)
      {:ok, state} = Fishing.apply_input(state, 0, %{"kind" => "cast", "power" => 0.7})
      {:ok, state} = Fishing.apply_input(state, 1, %{"kind" => "cast", "power" => 0.55})
      {state, outcome} = Fishing.step_simulation(state, %{}, 24)

      assert outcome == nil
      assert state["anglers"]["0"]["phase"] == "waiting"
      assert state["anglers"]["1"]["phase"] == "waiting"
      assert is_list(state["anglers"]["0"]["line"])
      assert is_list(state["anglers"]["1"]["bobber"])
    end
  end

  describe "live environment" do
    test "rain raises bite chance versus dry defaults" do
      dry = Fishing.bite_chance(%{"rain" => 0.0, "timePhase" => 0.5})
      wet = Fishing.bite_chance(%{"rain" => 0.9, "timePhase" => 0.5})
      assert wet > dry
    end

    test "identical live weather + seed + ticks reproduce the same bite tick" do
      env = %{"rain" => 0.8, "timePhase" => 0.15, "wind" => [0.0, 0.0]}

      bite_tick = fn ->
        state = Fishing.init_sim_state(slots: [0], environment: env, seed: 99, now_ms: 5000)
        {:ok, state} = Fishing.apply_input(state, 0, %{"kind" => "cast", "power" => 0.6})

        Enum.reduce_while(1..4000, state, fn _, acc ->
          {next, _} = Fishing.step_simulation(acc, %{}, 1)

          if next["anglers"]["0"]["phase"] == "bite" do
            {:halt, next["tickCount"]}
          else
            {:cont, next}
          end
        end)
      end

      a = bite_tick.()
      b = bite_tick.()
      assert is_integer(a)
      assert a == b
    end

    test "reel and leave are independent; catch is not inventory" do
      env = %{"rain" => 1.0, "timePhase" => 0.15, "wind" => [0.0, 0.0]}
      state = Fishing.init_sim_state(slots: [0, 1], environment: env, seed: 3)
      {:ok, state} = Fishing.apply_input(state, 0, %{"kind" => "cast", "power" => 0.8})
      {:ok, state} = Fishing.apply_input(state, 1, %{"kind" => "cast", "power" => 0.8})

      state =
        Enum.reduce_while(1..5000, state, fn _, acc ->
          {next, _} = Fishing.step_simulation(acc, %{}, 1)

          if next["anglers"]["0"]["phase"] == "bite" do
            {:halt, next}
          else
            {:cont, next}
          end
        end)

      assert state["anglers"]["0"]["phase"] == "bite"
      other = state["anglers"]["1"]["phase"]
      assert other in ["waiting", "bite"]

      {:ok, state} = Fishing.apply_input(state, 0, %{"kind" => "reel"})
      hooked = state["anglers"]["0"]["lastCatch"]
      assert is_binary(hooked["species"])
      refute Map.has_key?(hooked, "gold")
      refute Map.has_key?(hooked, "inventory")
      assert state["anglers"]["1"]["phase"] == other

      {state, _} = Fishing.step_simulation(state, %{}, 90)
      assert state["anglers"]["0"]["phase"] == "catch"

      {:ok, state} = Fishing.apply_input(state, 0, %{"kind" => "release"})
      assert state["anglers"]["0"]["phase"] == "idle"
      assert hd(state["recentReleases"])["species"] == hooked["species"]
      assert state["anglers"]["1"]["phase"] == other

      {:ok, state} = Fishing.apply_input(state, 1, %{"kind" => "leave"})
      assert state["anglers"]["1"]["phase"] == "idle"
      assert state["anglers"]["1"]["bobber"] == nil
      assert state["anglers"]["0"]["phase"] == "idle"
    end
  end
end

File: tests/fishing.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  FISHING_RULES_VERSION,
  FISH_SPECIES,
  DEFAULT_FISHING_ENVIRONMENT,
  initFishingState,
  validateFishingInput,
  applyFishingInput,
  applyFishingEnvironment,
  clearAngler,
  stepFishingSimulation,
  visibleLines,
  biteChancePerTick,
  pickSpecies,
} from '../shared/fishingModel.js';

test('fishing: init has no economy fields and starts both anglers idle', () => {
  const state = initFishingState({ activeSlots: [0, 1], nowMs: 1_700_000_000_000 });
  assert.equal(state.rulesVersion, FISHING_RULES_VERSION);
  assert.equal(state.status, 'open');
  assert.equal(state.environment.policy, 'live');
  assert.equal(state.environmentAt, 1_700_000_000_000);
  assert.equal(state.anglers[0].phase, 'idle');
  assert.equal(state.anglers[1].phase, 'idle');
  const blob = JSON.stringify(state);
  assert.equal(blob.includes('"gold"'), false);
  assert.equal(blob.includes('"inventory"'), false);
  assert.equal(blob.includes('"xp"'), false);
  assert.equal(blob.includes('"crop"'), false);
});

test('fishing: validateFishingInput accepts cast/reel/release and rejects junk', () => {
  assert.equal(validateFishingInput(null).valid, false);
  assert.equal(validateFishingInput([]).valid, false);
  assert.equal(validateFishingInput({ kind: 'warp' }).valid, false);

  const cast = validateFishingInput({ kind: 'cast', power: 9 });
  assert.equal(cast.valid, true);
  assert.equal(cast.sanitized.power, 1);

  assert.equal(validateFishingInput({ kind: 'reel' }).valid, true);
  assert.equal(validateFishingInput({ kind: 'release' }).valid, true);
  assert.equal(validateFishingInput({ kind: 'leave' }).valid, true);
  assert.equal(validateFishingInput({ kind: 'neutral' }).valid, true);
});

test('fishing: two visitors each see both lines after independent casts', () => {
  let state = initFishingState({ activeSlots: [0, 1], seed: 7 });
  state = applyFishingInput(state, 0, { kind: 'cast', power: 0.7 }).simState;
  state = applyFishingInput(state, 1, { kind: 'cast', power: 0.55 }).simState;
  state = stepFishingSimulation(state, {}, FISHING_CAST_STEPS()).simState;

  const lines = visibleLines(state);
  assert.equal(lines.length, 2);
  assert.ok(lines.every((row) => row.bobber && row.line?.length === 2));
  assert.equal(state.anglers[0].phase, 'waiting');
  assert.equal(state.anglers[1].phase, 'waiting');
});

function FISHING_CAST_STEPS() {
  return 24;
}

test('fishing: live rain raises bite chance vs the declared dry default', () => {
  const dry = biteChancePerTick({ ...DEFAULT_FISHING_ENVIRONMENT, rain: 0, timePhase: 0.5 });
  const wet = biteChancePerTick({ ...DEFAULT_FISHING_ENVIRONMENT, rain: 0.9, timePhase: 0.5 });
  assert.ok(wet > dry, `rain should raise bite chance: ${wet} vs ${dry}`);
});

test('fishing: identical live environment + seed + ticks reproduce the same bite', () => {
  const env = { ...DEFAULT_FISHING_ENVIRONMENT, rain: 0.8, timePhase: 0.15 };
  const run = () => {
    let state = initFishingState({ activeSlots: [0], environment: env, seed: 99, nowMs: 5000 });
    state = applyFishingInput(state, 0, { kind: 'cast', power: 0.6 }).simState;
    let biteTick = null;
    for (let i = 0; i < 4000; i++) {
      const res = stepFishingSimulation(state, {}, 1);
      state = res.simState;
      if (res.events.some((e) => e.type === 'bite')) {
        biteTick = state.tickCount;
        break;
      }
    }
    return { biteTick, phase: state.anglers[0].phase };
  };

  const a = run();
  const b = run();
  assert.ok(a.biteTick !== null, 'expected a bite under heavy rain/dawn');
  assert.equal(a.biteTick, b.biteTick);
  assert.equal(a.phase, b.phase);
});

test('fishing: reel/leave are independent and catch/release writes no inventory', () => {
  const env = { ...DEFAULT_FISHING_ENVIRONMENT, rain: 1, timePhase: 0.15 };
  let state = initFishingState({ activeSlots: [0, 1], environment: env, seed: 3 });
  state = applyFishingInput(state, 0, { kind: 'cast', power: 0.8 }).simState;
  state = applyFishingInput(state, 1, { kind: 'cast', power: 0.8 }).simState;

  let hooked = false;
  for (let i = 0; i < 5000 && !hooked; i++) {
    const res = stepFishingSimulation(state, {}, 1);
    state = res.simState;
    if (state.anglers[0].phase === 'bite') hooked = true;
  }
  assert.equal(state.anglers[0].phase, 'bite');
  assert.ok(state.anglers[1].phase === 'waiting' || state.anglers[1].phase === 'bite');

  const reel = applyFishingInput(state, 0, { kind: 'reel' });
  state = reel.simState;
  assert.equal(reel.event?.type, 'fish_hooked');
  assert.ok(FISH_SPECIES.includes(state.anglers[0].lastCatch.species));
  assert.equal(Object.hasOwn(state.anglers[0].lastCatch, 'gold'), false);
  assert.equal(Object.hasOwn(state.anglers[0].lastCatch, 'inventory'), false);

  const stillWaiting = state.anglers[1].phase;
  assert.ok(stillWaiting === 'waiting' || stillWaiting === 'bite');

  state = stepFishingSimulation(state, {}, 90).simState;
  assert.equal(state.anglers[0].phase, 'catch');

  const released = applyFishingInput(state, 0, { kind: 'release' });
  state = released.simState;
  assert.equal(released.event?.type, 'fish_released');
  assert.equal(state.anglers[0].phase, 'idle');
  assert.equal(state.recentReleases[0].species, released.event.payload.species);
  assert.equal(state.anglers[1].phase, stillWaiting);

  const afterLeave = clearAngler(state, 1);
  assert.equal(afterLeave.anglers[1].phase, 'idle');
  assert.equal(afterLeave.anglers[1].bobber, null);
  assert.equal(afterLeave.anglers[0].phase, 'idle');
});

test('fishing: timestamped live weather updates change later bite odds', () => {
  let state = initFishingState({
    activeSlots: [0],
    environment: { ...DEFAULT_FISHING_ENVIRONMENT, rain: 0 },
    seed: 11,
    nowMs: 1000,
  });
  const dryChance = biteChancePerTick(state.environment);
  state = applyFishingEnvironment(state, { ...DEFAULT_FISHING_ENVIRONMENT, rain: 1, timePhase: 0.15 }, 2500);
  assert.equal(state.environmentAt, 2500);
  assert.ok(biteChancePerTick(state.environment) > dryChance);
  assert.ok(pickSpecies(state.environment, 11, 0, 40));
});

test('fishing: missing environment falls back to the declared shared default', () => {
  const state = initFishingState({ activeSlots: [0], environment: null });
  assert.equal(state.environment.policy, 'live');
  assert.equal(state.environment.rain, DEFAULT_FISHING_ENVIRONMENT.rain);
  assert.deepEqual(state.environment.wind, [0, 0]);
});
